### PR TITLE
feat(annotations): authorization service

### LINKED ISSUE #21214: Implement authorization & access controls (annotations)
This will entail mechanisms to ensure that users can only performed actions on annotations that they are authorized to do.

### PR DESCRIPTION
**Pull request by wbaker85**
Closes #21214

This is the authorization service implementation for annotations and streams. Both annotations and streams will use the same "annotations" resource type since streams can only be interacted with via annotations.

As I developed the authorization code, I found a few adjustments that needed to be made to the service definition, which I have highlighted in the review notes.

### CODE REVIEW

File: annotation.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ type AnnotationService interface {
 	// ListAnnotations lists all annotations matching the filter.
 	ListAnnotations(ctx context.Context, orgID platform.ID, filter AnnotationListFilter) (ReadAnnotations, error)
 	// GetAnnotation gets an annotation by id.
-	GetAnnotation(ctx context.Context, id platform.ID) (*AnnotationEvent, error)
+	GetAnnotation(ctx context.Context, id platform.ID) (*StoredAnnotation, error)
```
Comment (wbaker85) [inline annotation.go:73]: For `GetAnnotation` and `GetStream` (below), we need to return the full DB representation of the resource, so that the orgID is available. This is so that the authorization service can verify that the request has read access for that organization.

File: annotations/transport/annotations_router.go [hunk 2]
```diff
@@ -106,13 +107,19 @@ func (h *AnnotationHandler) handleGetAnnotation(w http.ResponseWriter, r *http.R
...
 
-	a, err := h.annotationService.GetAnnotation(ctx, *id)
+	s, err := h.annotationService.GetAnnotation(ctx, *id)
 	if err != nil {
 		h.api.Err(w, r, err)
 		return
 	}
 
-	h.api.Respond(w, r, http.StatusOK, a)
+	c, err := storedAnnotationToEvent(s)
```
Comment (wbaker85) [inline annotations/transport/annotations_router.go:116]: Per the service changes above, the handler needs to convert the `StoredAnnotation` type to an `AnnotationEvent` before responding per the API contract.

File: annotations/transport/annotations_router.go [hunk 3]
```diff
@@ -234,3 +241,42 @@ func decodeUpdateAnnotationRequest(r *http.Request) (*influxdb.AnnotationCreate,
 
 	return u, nil
 }
+
+func storedAnnotationToEvent(s *influxdb.StoredAnnotation) (*influxdb.AnnotationEvent, error) {
```
Comment (wbaker85) [inline annotations/transport/annotations_router.go:245]: These two functions for converting stored annotations to events are here for now, but I'll probably move them when I make the storage service implementation, since there will be other similar functions needed there.

File: annotation.go [hunk 4]
```diff
@@ -81,11 +81,14 @@ type AnnotationService interface {
 	// ListStreams lists all streams matching the filter.
 	ListStreams(ctx context.Context, orgID platform.ID, filter StreamListFilter) ([]ReadStream, error)
 	// CreateOrUpdateStream creates or updates the matching stream by name.
-	CreateOrUpdateStream(ctx context.Context, stream Stream) (*ReadStream, error)
+	CreateOrUpdateStream(ctx context.Context, orgID platform.ID, stream Stream) (*ReadStream, error)
```
Comment (wbaker85) [inline annotation.go:84]: Needs the `orgID` here for the case of creating a stream - the backend needs to know what org to create the stream for.

File: annotation.go [hunk 5]
```diff
@@ -81,11 +81,14 @@ type AnnotationService interface {
...
+	// GetStream gets a stream by id. Currently this is only used for authorization, and there are no
+	// API routes for getting a single stream by ID.
+	GetStream(ctx context.Context, id platform.ID) (*StoredStream, error)
 	// UpdateStream updates the stream by the ID.
 	UpdateStream(ctx context.Context, id platform.ID, stream Stream) (*ReadStream, error)
 	// DeleteStreams deletes one or more streams by name.
-	DeleteStreams(ctx context.Context, delete BasicStream) error
+	DeleteStreams(ctx context.Context, orgID platform.ID, delete BasicStream) error
```
Comment (wbaker85) [inline annotation.go:91]: Needs the `orgID` here as well since the backend needs to know which orgs to delete streams for.

File: authorizer/notebook_test.go [hunk 6]
```diff
@@ -61,7 +61,7 @@ func Test_GetNotebook(t *testing.T) {
 				GetNotebook(gomock.Any(), *nbID).
 				Return(newTestNotebook(*orgID1), nil)
 
-			perm := newTestPermission(influxdb.ReadAction, tt.permissionOrg)
+			perm := newTestNotebooksPermission(influxdb.ReadAction, tt.permissionOrg)
```
Comment (wbaker85) [inline authorizer/notebook_test.go:64]: Name changes in this file were to disambiguate in the `authorizer_test` package.

File: authz_test.go [hunk 7]
```diff
@@ -325,7 +325,8 @@ func TestPermissionAllResources_Valid(t *testing.T) {
 		platform.BucketsResourceType,
 		platform.DashboardsResourceType,
 		platform.SourcesResourceType,
-		platform.DashboardsResourceType,
+		platform.NotebooksResourceType,
```
Comment (wbaker85) [inline authz_test.go:328]: Fixes a copy/paste error I made in #21490 😨

File: annotations/transport/annotations_router.go [hunk 8]
```diff
@@ -234,3 +241,42 @@ func decodeUpdateAnnotationRequest(r *http.Request) (*influxdb.AnnotationCreate,
...
+
+}
+
+func stickerSliceToMap(stickers []string) map[string]string {
+	stickerMap := map[string]string{}
+
+	for i := range stickers {
+		sticks := strings.SplitN(stickers[i], "=", 2)
+		if len(sticks) < 2 {
+			continue
```
Comment (danxmoran) [inline annotations/transport/annotations_router.go:276]: Should we make this an error case, or log it? I worry about silently ignoring entries if they're malformatted will cause headaches later.

File: authorizer/annotation.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+	if _, _, err := AuthorizeCreate(ctx, influxdb.AnnotationsResourceType, orgID); err != nil {
+		return nil, err
+	}
+
+	return s.s.CreateAnnotations(ctx, orgID, create)
+}
+
+// ListAnnotations checks to see if the authorizer on context has read access for annotations for the provided orgID
+func (s *AnnotationService) ListAnnotations(ctx context.Context, orgID platform.ID, filter influxdb.AnnotationListFilter) (influxdb.ReadAnnotations, error) {
+	if _, _, err := AuthorizeOrgReadResource(ctx, influxdb.AnnotationsResourceType, orgID); err != nil {
```
Comment (danxmoran) [inline authorizer/annotation.go:36]: For other resources (i.e. buckets) we instead run the underlying `List` operation, then filter the results based on auth. Should we do the same here for consistency?

File: authorizer/annotation.go [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	if _, _, err := AuthorizeWrite(ctx, influxdb.AnnotationsResourceType, id, a.OrgID); err != nil {
+		return nil, err
+	}
+	return s.s.UpdateAnnotation(ctx, id, update)
+}
+
+// ListStreams checks to see if the authorizer on context has read access for streams for the provided orgID
+func (s *AnnotationService) ListStreams(ctx context.Context, orgID platform.ID, filter influxdb.StreamListFilter) ([]influxdb.ReadStream, error) {
+	if _, _, err := AuthorizeOrgReadResource(ctx, influxdb.AnnotationsResourceType, orgID); err != nil {
```
Comment (danxmoran) [inline authorizer/annotation.go:89]: Same question here.

File: authorizer/annotation.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+	// note: a given org can only have one stream by the same name. this constraint is enforced in the database schema
+	streams, err := s.s.ListStreams(ctx, orgID, influxdb.StreamListFilter{
+		StreamIncludes: []string{stream.Name},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// for updating an existing stream, we need to get the stream ID before calling the UpdateStream method
+	if len(streams) != 0 {
```
Comment (danxmoran) [inline authorizer/annotation.go:122]: Could we instead do:
```suggestion
	if len(streams) == 1 {
```
Even though I read the comment above about only having 1 stream per org with a given name, my brain still reacted like "what about the rest of the `streams`?" when I saw `streams[o].ID` below.

File: authorizer/annotation.go [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+		st, err := s.s.GetStream(ctx, streams[0].ID)
+		if err != nil {
+			return nil, err
+		}
+
```
Comment (danxmoran) [inline authorizer/annotation.go:127]: I think there's a missing auth-check here? Should be checking write-permission between the Get and the Update

Comment (danxmoran) [reply authorizer/annotation.go:127]: Alt: it looks like the body here is almost the same as `UpdateStream` below, could we replace the whole thing with:
```
return s.UpdateStream(ctx, streams[0].ID, stream)
```
?

File: authz.go [hunk 13]
```diff
@@ -137,6 +137,8 @@ const (
 	DBRPResourceType = ResourceType("dbrp") // 17
 	// NotebooksResourceType gives permission to one or more notebooks.
 	NotebooksResourceType = ResourceType("notebooks") // 18
+	// AnnotationsResourceType gives permission to one or more annotations.
+	AnnotationsResourceType = ResourceType("annotations") // 18
```
Comment (danxmoran) [inline authz.go:141]:
```suggestion
	AnnotationsResourceType = ResourceType("annotations") // 19
```

File: authz.go [hunk 14]
```diff
@@ -179,6 +182,7 @@ var OrgResourceTypes = []ResourceType{
 	ChecksResourceType,               // 16
 	DBRPResourceType,                 // 17
 	NotebooksResourceType,            // 18
+	AnnotationsResourceType,          // 18
```
Comment (danxmoran) [inline authz.go:185]:
```suggestion
	AnnotationsResourceType,          // 19
```

File: authz.go [hunk 15]
```diff
@@ -208,6 +212,7 @@ func (t ResourceType) Valid() (err error) {
 	case ChecksResourceType: // 16
 	case DBRPResourceType: // 17
 	case NotebooksResourceType: // 18
+	case AnnotationsResourceType: // 18
```
Comment (danxmoran) [inline authz.go:215]:
```suggestion
	case AnnotationsResourceType: // 19
```

File: annotations/transport/annotations_router.go (hunk 8 above)
Comment (wbaker85) [reply annotations/transport/annotations_router.go]: I basically just lifted this code from IDPE, but I do agree about silently ignoring errors being bad so I went ahead and made it return and error and added some more tests cases for it.

File: authorizer/annotation.go (hunk 11 above)
Comment (wbaker85) [reply authorizer/annotation.go]: Seems like another reasonable edge case to guard against, so I updated the code to do specific things on `==1` or `==0`, and return a server error if more than one matching stream is returned somehow.

File: authorizer/annotation.go (hunk 12 above)
Comment (wbaker85) [reply authorizer/annotation.go:133]: Oh yeah good catch...I think I meant to return `s.UpdateStream(...)` instead of `s.s.UpdateStream(...)` here anyway. But yeah that's pretty redundant so I made the update like you suggested! It helped simplify that part quite a bit.

File: authorizer/annotation.go (hunk 9 above)
Comment (wbaker85) [reply authorizer/annotation.go:38]: Yeah this is interesting. Fundamentally I think that if you request a list of a resource for an org that you don't have read access to, you should get a forbidden response rather than an empty list. Otherwise there's no way to tell if you are requesting a list for an org that doesn't have any resources, or if you just aren't authenticated right. But indeed that is not consistent with how other resources work.

Trying to leverage as much of the IDPE code as I could, there wouldn't immediately be a way to get the entire list and then filter it by auth like the other resources, since the `ReadStream` and `ReadAnnotation` structs don't have the `orgID` in them - although I've already had to make some changes to responses and structs elsewhere so it's not a total show stopper.

Beyond my own personal biases and laziness, I think having it like this is going to be more consistent with the cloud API, although it's hard to say exactly how that will work since it doesn't have multi-org yet. I know that as it is now, the annotations service on cloud would give a forbidden response if you do a GET request for an org you don't have read access to, even if that's not really even possible to do.

So that's my reasoning for setting it up this way, mostly leaning into the consistency with cloud aspect, although I feel like there might be a gap with how other resources types respond on OSS vs. cloud. Annotations (and notebooks) are at least behind the `/api/v2private` endpoint so maybe there's some room for them to behave differently than other resources on OSS. Let me know what you think!

Comment (danxmoran) [reply authorizer/annotation.go:38]: Is the annotations permission an all-or-nothing within a single org? Or can users be granted read-permission on some-but-not-all of the annotations? If fine-grained auth is possible then I think we _have_ to post-filter the list results. I do like an explicit 403 on bad auth + consistency with cloud, so we could leave this top-level check in place either way.

Comment (wbaker85) [reply authorizer/annotation.go:39]: There's not fine-grained auth currently but who knows what the future holds. So I kept the top-level check and added the secondary filter! Seems like this should be good to keep things consistent and set us up for success in the future if fine-grained auth of things like annotations becomes necessary.

**Review by danxmoran — APPROVED** [review]
